travis/resource_key_pair.go

```diff
@@ -0,0 +1,179 @@
+				Description:  "Same as {repository.owner.name}/{repository.name}.",
+				ForceNew:     true,
+				ExactlyOneOf: []string{"repository_id"},
+			},
+			"description": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "A text description of this key pair.",
+			},
+			"value": &schema.Schema{
+				Type:        schema.TypeString,
+				Description: "The private key",
+				Required:    true,
+				Sensitive:   true,
+			},
+			"fingerprint": &schema.Schema{
+				Type:        schema.TypeString,
+				Description: "Fingerprint of the RSA key",
+				Computed:    true,
+			},
+			"public_key": &schema.Schema{
+				Type:        schema.TypeString,
+				Description: "The public key.",
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func resourceKeyPairCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var (
+		client  = m.(*Client)
+		keyPair *travis.KeyPair
+		err     error
+	)
+	if repoID, ok := d.GetOk("repository_id"); ok {
+		keyPair, _, err = client.KeyPair.CreateByRepoId(ctx, repoID.(uint), generateKeyPairBody(d))
+		if err != nil {
+			return diag.Errorf("error creating key pair by repo ID (%d): %s", repoID, err)
+		}
+	} else if repoSlug, ok := d.GetOk("repository_slug"); ok {
+		keyPair, _, err = client.KeyPair.CreateByRepoSlug(ctx, repoSlug.(string), generateKeyPairBody(d))
+		if err != nil {
+			return diag.Errorf("error creating key pair by repo slug (%s): %s", repoSlug, err)
+		}
+	} else {
+		return diag.Errorf("one of repository_id or repository_slug must be specified")
+	}
+	assignKeyPair(keyPair, d)
+	return nil
+}
+
+func resourceKeyPairRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var (
+		client  = m.(*Client)
+		keyPair *travis.KeyPair
+		err     error
+	)
+	if repoID, ok := d.GetOk("repository_id"); ok {
+		keyPair, _, err = client.KeyPair.FindByRepoId(ctx, repoID.(uint))
+		if err != nil {
+			if isNotFound(err) {
+				d.SetId("")
+				return nil
+			}
+			return diag.Errorf("error reading key pair by repo ID (%d): %s", repoID, err)
+		}
+	} else if repoSlug, ok := d.GetOk("repository_slug"); ok {
+		keyPair, _, err = client.KeyPair.FindByRepoSlug(ctx, repoSlug.(string))
+		if err != nil {
+			if isNotFound(err) {
+				d.SetId("")
+				return nil
+			}
+			return diag.Errorf("error reading key pair by repo slug (%s): %s", repoSlug, err)
+		}
+	} else {
+		return diag.Errorf("one of repository_id or repository_slug must be specified")
+	}
+	assignKeyPair(keyPair, d)
+	return nil
+}
+
+func resourceKeyPairUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var (
+		client = m.(*Client)
+		err    error
+		update = &travis.KeyPairBody{}
+	)
+
+	if d.HasChange("value") {
+		update.Value = d.Get("value").(string)
+	}
+
+	if d.HasChange("description") {
+		update.Description = d.Get("description").(string)
+	}
+
+	if update.Value != "" || update.Description != "" {
```

Comment:
I seem Travis accepts an empty description.

```suggestion
	if update.Value != "" {
```

Reply:
Wouldn't this mean that the description would only be updated if you also updated the value at the same time?

I did a little testing and the Travis API does accept an empty description on both creation and update. The client library here won't update the description to an empty string from what I can tell.

```go
package main

import (
	"context"
        "fmt"
	"github.com/shuheiktgw/go-travis"
)

func main() {
	client := travis.NewClient(...)
	repoSlug := "repo slug"

	keyPair, _, _ := client.KeyPair.FindByRepoSlug(context.Background(), repoSlug)
	fmt.Printf("Existing keypair description: %s\n", *keyPair.Description)

	update := &travis.KeyPairBody{
		Description: "",
	}
	keyPair, _, _ = client.KeyPair.UpdateByRepoSlug(context.Background(), repoSlug, update)
	fmt.Printf("Keypair description returned by PATCH: %s\n", *keyPair.Description)

	keyPair, _, _ = client.KeyPair.FindByRepoSlug(context.Background(), repoSlug)
	fmt.Printf("Keypair desc fetched again: %s\n", *keyPair.Description)
}
```
If I run this against one of my repos I get this:

```
Existing keypair description: Service Key
Keypair description returned by PATCH: Service Key
Keypair desc fetched again: Service Key
```

I was able to use curl to set an empty description on that same keypair however.